src/renderer/dx/CustomTextLayout.cpp

```diff
@@ -303,6 +341,45 @@ CATCH_RETURN()
             _glyphIndices.resize(totalGlyphsArrayCount);
         }
 
+        if (_isEntireTextSimple)
+        {
+            // When the entire text is simple, we can skip GetGlyphs and directly retrieve glyph indices and
+            // advances(in font design unit). With the help of font metrics, we can calculate the actual glyph
+            // advances without the need of GetGlyphPlacements. This shortcut will significantly reduce the time
+            // needed for text analysis.
+            DWRITE_FONT_METRICS1 metrics;
+            run.fontFace->GetMetrics(&metrics);
+
+            // With simple text, there's only one run. The actual glyph count is the same as textLength.
+            _glyphDesignUnitAdvances.resize(textLength);
+            _glyphAdvances.resize(textLength);
+
+            USHORT designUnitsPerEm = metrics.designUnitsPerEm;
+
+            RETURN_IF_FAILED(_font->GetDesignGlyphAdvances(
+                textLength,
+                &_glyphIndices.at(glyphStart),
+                &_glyphDesignUnitAdvances.at(glyphStart),
+                run.isSideways));
+
+            for (size_t i = glyphStart; i < _glyphAdvances.size(); i++)
+            {
+                _glyphAdvances.at(i) = (float)_glyphDesignUnitAdvances.at(i) / designUnitsPerEm * _format->GetFontSize() * run.fontScale;
+            }
+
+            // Empty all the offsets, we're not going to use them
+            _glyphOffsets.assign(textLength, {});
+
+            // Set all the clusters as sequential. In a simple run, we're going 1 to 1.
+            // Fill the clusters sequentially from 0 to N-1.
+            std::iota(_glyphClusters.begin(), _glyphClusters.end(), gsl::narrow_cast<unsigned short>(0));
+
+            run.glyphCount = textLength;
+            glyphStart += textLength;

```

Comment:
I'm a bit concerned about these three loops. One of them goes from `glyphStart` to `end`, and the other two touch _the entire collection_. Are the collections guaranteed to be the exact right length, and not have any values before `glyphStart`?

Regardless, is it possible to make these three loops just, like, one loop?

Reply:
(three loops = `for`, `assign`, `iota`)

Reply:
I cut the assign one as a part of other stuff. Now it's just two and I think the `iota` is elegant. So maybe leave it?